swig/include/python/gdal_python.i

```diff
@@ -5488,6 +5605,111 @@ class VSIFile(BytesIO):
 %extend GDALAlgorithmHS {
 %pythoncode %{
 
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        if hasattr(self, "has_run") and not self.Finalize():
+            # We go here only if gdal.UseExceptions() has not been called
+            raise RuntimeError("Algorithm.Finalize() failed: %s" % GetLastErrorMsg())
+
+    def _get_arg_value(self, arg, parse_json, dataset_instance):
+        val = arg.Get()
+        if parse_json and arg.GetType() == GAAT_STRING and \
+           ((val.startswith('{') and (val.endswith('}') or val.endswith('}\n'))) or \
+           (val.startswith('[') and (val.endswith(']') or val.endswith(']\n')))):
+            import json
+            try:
+                return json.loads(val)
+            except Exception:
+                return val
+        elif dataset_instance and arg.GetType() == GAAT_DATASET:
+            return val.GetDataset()
+        else:
+            return val
+
+
+    def output(self, parse_json=True, dataset_instance=True):
+        """Return the single output value of this algorithm, after it has been run.
+
+           If there are multiple output values, this method cannot be called,
+           and the :py:meth:`outputs` (plural) method should be called instead.
+
+           Parameters
+           -----------
+           parse_json: bool
+               Whether a JSON string should be returned as a dict or list (instead of a string).
+               Defaults to True
+           dataset_instance: bool
+               Whether arguments of type GAAT_DATASET should be returned as a Dataset instance
+               using the :py:meth:`gdal.AlgorithmArg.GetDataset` method rather than a
+               gdal.AlgorithmArg instance.
```

Comment:
Is there a case where a user would want to get an `AlgoritmArg` returned (instead of a Dataset or any other type)?

Reply:
> Is there a case where a user would want to get an `AlgoritmArg` returned (instead of a Dataset or any other type)?

for an hypothetic algorithm that would *generate* a dataset name. But that seems like premature complication. I'll remove that parameter. If the hypothetical use case arises we can extend the API